client/file-manager/src/rocksdb.rs

```diff
@@ -0,0 +1,925 @@
+        // Need to clone because we cannot have a immutable borrow after mutably borrowing
+        // in the next step.
+        let trie_root_key = root;
+        let mut trie = TrieDBMutBuilder::<T>::from_existing(db, &mut root).build();
+
+        for i in 0..chunk_count {
+            trie.remove(&ChunkId::new(i as u64).as_trie_key())
+                .map_err(|e| {
+                    error!(target: LOG_TARGET, "Failed to delete chunk from RocksDb: {}", e);
+                    FileStorageError::FailedToWriteToStorage
+                })?;
+        }
+
+        // Remove the root from the trie.
+        trie.remove(trie_root_key.as_ref()).map_err(|e| {
+            error!(target: LOG_TARGET, "Failed to delete root from RocksDb: {}", e);
+            FileStorageError::FailedToWriteToStorage
+        })?;
+
+        let new_root = *trie.root();
+
+        drop(trie);
+
+        // TODO: improve error handling
+        // Commit the changes to disk.
+        self.commit(trie_root_key).map_err(|e| {
+            error!(target: LOG_TARGET, "Failed to commit changes to persistent storage: {}", e);
+            FileStorageError::FailedToWriteToStorage
+        })?;
+
+        // Set new internal root (no trie)
+        self.root = new_root;
+
+        self.chunk_count = 0;
+
+        Ok(())
+    }
+}
+
+impl<T: TrieLayout + Send + Sync> AsHashDB<HashT<T>, DBValue> for RocksDbFileDataTrie<T>
+where
+    HasherOutT<T>: TryFrom<[u8; H_LENGTH]>,
+{
+    fn as_hash_db<'b>(&'b self) -> &'b (dyn HashDB<HashT<T>, DBValue> + 'b) {
+        self
+    }
+    fn as_hash_db_mut<'b>(&'b mut self) -> &'b mut (dyn HashDB<HashT<T>, DBValue> + 'b) {
+        &mut *self
+    }
+}
+
+impl<T: TrieLayout + Send + Sync> hash_db::HashDB<HashT<T>, DBValue> for RocksDbFileDataTrie<T>
+where
+    HasherOutT<T>: TryFrom<[u8; H_LENGTH]>,
+{
+    fn get(&self, key: &HasherOutT<T>, prefix: Prefix) -> Option<DBValue> {
+        HashDB::get(&self.overlay, key, prefix).or_else(|| {
+            self.storage.get(key, prefix).unwrap_or_else(|e| {
+                warn!(target: LOG_TARGET, "Failed to read from DB: {}", e);
+                None
+            })
+        })
+    }
+
+    fn contains(&self, key: &HasherOutT<T>, prefix: Prefix) -> bool {
+        HashDB::contains(&self.overlay, key, prefix)
+    }
+
+    fn insert(&mut self, prefix: Prefix, value: &[u8]) -> HasherOutT<T> {
+        HashDB::insert(&mut self.overlay, prefix, value)
+    }
+
+    fn emplace(&mut self, key: HasherOutT<T>, prefix: Prefix, value: DBValue) {
+        HashDB::emplace(&mut self.overlay, key, prefix, value)
+    }
+
+    fn remove(&mut self, key: &HasherOutT<T>, prefix: Prefix) {
+        HashDB::remove(&mut self.overlay, key, prefix)
+    }
+}
+
+pub struct RocksDbFileStorage<T: TrieLayout + 'static>
+where
+    HasherOutT<T>: TryFrom<[u8; H_LENGTH]>,
+{
+    pub metadata: HashMap<HasherOutT<T>, FileMetadata>,
+    pub file_data: HashMap<HasherOutT<T>, RocksDbFileDataTrie<T>>,
```

Comment:
I am a little confused here. All this data should be persistently stored and the RocksDbFileStorage shouldn't have to rely on ephemeral data to track files. 

The functions should all be accepting identifiers, roots, etc. which you can then query from your storage. I am wondering why we even need `RocksDbFileStorage`. It doesn't serve much purpose.

Reply:
This is now fixed with a pointer to a persistent storage inside `FileStorage` :-)